lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        key = Rails.application.key_generator.generate_key("discourse-auth-cookie", 32)
+        ActiveSupport::MessageEncryptor.new(
+          key,
+          key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
+        )
+      end
+    end
+  end
+
+  TOKEN_SIZE ||= 32
+
+  TOKEN_KEY ||= "token"
+  ID_KEY ||= "id"
+  TL_KEY ||= "tl"
+  TIME_KEY ||= "time"
+  VALID_KEY ||= "valid"
+  private_constant *%i[
+    TOKEN_KEY
+    ID_KEY
+    TL_KEY
+    TIME_KEY
+    VALID_KEY
+  ]
+
+  attr_reader *%i[token user_id trust_level timestamp valid_for]
```

Comment:
```suggestion
  attr_reader :token, :user_id, :trust_level, :timestamp, :valid_for
```
Using an array and then spreading it seems unnecessary to me here.